File: backend/src/mappers/quiz.ts
import type { Quiz } from 'vulms-sdk';
import type { QuizDto, QuizSummaryDto } from '../types/quizzes.js';

function toDateStr(d: Date | string | undefined): string | undefined {
  if (!d) return undefined;
  if (d instanceof Date) return d.toISOString();
  return d;
}

export function toQuizDto(q: Quiz): QuizDto {
  return {
    courseCode: q.courseCode,
    courseTitle: q.courseTitle,
    title: q.title,
    startDate: toDateStr(q.startDate),
    endDate: toDateStr(q.endDate),
    totalMarks: q.totalMarks,
    obtainedMarks: q.obtainedMarks,
    availabilityStatus: q.availabilityStatus,
    submissionStatus: q.submissionStatus,
    resultStatus: q.resultStatus,
    submitDate: toDateStr(q.submitDate),
  };
}

export function toQuizSummaryDto(summary: {
  total: number;
  submitted: number;
  pending: number;
  open: number;
  closed: number;
  resultDeclared: number;
}): QuizSummaryDto {
  return {
    total: summary.total,
    submitted: summary.submitted,
    pending: summary.pending,
    open: summary.open,
    closed: summary.closed,
    resultDeclared: summary.resultDeclared,
  };
}

File: vulms_dart/lib/src/models/quiz.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'quiz.freezed.dart';
part 'quiz.g.dart';

/// Availability status of a quiz.
enum QuizAvailabilityStatus { open, closed, upcoming, unknown }

/// Submission status of a quiz.
enum QuizSubmissionStatus { submitted, notSubmitted, unknown }

/// Result status of a quiz.
enum QuizResultStatus { declared, pending, unknown }

/// Represents a VULMS quiz.
@freezed
abstract class Quiz with _$Quiz {
  const factory Quiz({
    required String courseCode,
    required String courseTitle,
    required String title,
    DateTime? startDate,
    DateTime? endDate,
    double? totalMarks,
    double? obtainedMarks,
    @Default(QuizAvailabilityStatus.unknown)
    QuizAvailabilityStatus availabilityStatus,
    @Default(QuizSubmissionStatus.unknown) QuizSubmissionStatus submissionStatus,
    @Default(QuizResultStatus.unknown) QuizResultStatus resultStatus,
    DateTime? submitDate,
  }) = _Quiz;

  factory Quiz.fromJson(Map<String, dynamic> json) => _$QuizFromJson(json);
}
